lib/Dialect/Secret/Transforms/DistributeGeneric.cpp

```diff
@@ -229,66 +237,228 @@ struct SplitGeneric : public OpRewritePattern<GenericOp> {
     // RegionBranchOpInterface (scf.while, scf.if).
   }
 
-  void splitGenericAroundOp(GenericOp op, Operation &opToDistribute,
-                            PatternRewriter &rewriter) const {
-    assert(false && "not implemented");
+  // Adds new value to yield in a generic op. This requires replacing the entire
+  // generic op and cloning all its ops, since I can't find a way to modify the
+  // op's return type in-place.
+  //
+  // Returns the new op and the value range corresponding to the new result
+  // values of the generic.
+  std::pair<GenericOp, ValueRange> addNewYieldedValues(
+      GenericOp genericOp, ValueRange newValuesToYield,
+      PatternRewriter &rewriter) const {
+    YieldOp yieldOp = genericOp.getYieldOp();
+    LLVM_DEBUG(yieldOp.emitRemark() << "Before adding new values\n");
+    yieldOp.getValuesMutable().append(newValuesToYield);
+    // Can't print here because yield will fail verification
+    auto newTypes = llvm::to_vector<4>(
+        llvm::map_range(yieldOp.getValues().getTypes(), [](Type t) -> Type {
+          SecretType newTy = secret::SecretType::get(t);
+          LLVM_DEBUG(llvm::dbgs() << "Adding new type: " << newTy << "\n");
```

Comment:
I'm surprised this works for a yield that already has some secret values yielded. wouldn't it double-secret the yielded type? (it's not clear to me that it's tested - both testcases are empty yields)

Reply:
The `yield` operand is always plaintext, and this `newTypes` variable is passed to `rewriter.create` as the result `TypeRange` of the overall `generic` op.

Reply:
ah right! thank you!